Add Go support

kazk opened the pull request:
Adding [Go](https://golang.org/) support (#5, #261). Uses [Ginkgo](http://onsi.github.io/ginkgo/) with [Gomega](http://onsi.github.io/gomega/).

---

Additional package candidates. Need feedbacks from Go users.

## Web Frameworks

- [Revel](https://revel.github.io/)
- [Beego](https://beego.me/)

## Database

- sqlite3
  - Package sqlite3 provides interface to SQLite3 databases.
- redis
  - Package redis is a client for the Redis database.
- pq
  - Package pq is a pure Go Postgres driver for the database/sql package.

## Testing

- [testify](https://github.com/stretchr/testify)
  - [assert](https://github.com/stretchr/testify#assert-package)
    - Package assert provides a set of comprehensive testing tools for use with the normal Go testing system.
  - [mock](https://github.com/stretchr/testify#mock-package)
    - Package mock provides a system by which it is possible to mock your objects and verify calls are happening as expected.

- [check](http://labix.org/gocheck)
  - Package check is a rich testing extension for Go's testing package.

- [ghttp](http://onsi.github.io/gomega/#ghttp-testing-http-clients)
  - Package ghttp supports testing HTTP clients by providing a test server (simply a thin wrapper around httptest's server) that supports registering multiple handlers.

kazk commented:
- [ ] Write Ginkgo custom reporter `CodewarsReporter`
- [ ] Decide how to structure tests
  - temporary file locations
  - putting test in `solution` package or separate `solution_test` package
  - wrap `opts.fixture` in "lib/runners/go.js" instead of making user import Ginkgo?
- [ ] Decide which Go version to add. v1.6/v1.7/v1.8
- [ ] Additional packages?

jhoffner commented on frameworks/go/codewars/reporter/reporter.go:
```diff
@@ -0,0 +1,87 @@
+  s = strings.Replace(s, "&", "&amp;", -1)
+  s = strings.Replace(s, "\"", "&quot;", -1)
+  s = strings.Replace(s, "'", "&#39;", -1)
+  s = strings.Replace(s, "<", "&lt;", -1)
```
No need to escape HTML characters

kazk replied:
Thanks for reviewing! I'll remove those lines.

---

For the version, I will try 1.7 for now because 1.8 isn't released yet. Maybe in few weeks I can upgrade to 1.8.
> Go 1.8 is not yet released. These are work-in-progress release notes. Go 1.8 is expected to be released in February 2017.
> <https://beta.golang.org/doc/go1.8>

---

> As close to how you would normally structure an app is best.

I'll look into this.

kazk commented:
@jhoffner what is the best way to clean up after the test? I'm trying to call some function after the test in `lib/runners/go.js`. I don't think I can use `temp` module because I need a folder in specific path with specific name.

I'm unable to pass `{onCompleted: [cleanup]}` to `shovel.start` because it empties that `opts.onCompleted = []` ([shovel.js#L40](https://github.com/Codewars/codewars-runner-cli/blob/master/lib/shovel.js#L40)).

```javascript
    // in case anything needs to cleanup after itself
    opts.onCompleted = [];
    // did you mean to do this?
    // if (!Array.isArray(opts.onCompleted)) opts.onCompleted = [];

    cb = cb || function (){};
    run(opts, strategies, function (buffer)
    {
        reportBuffer(opts, buffer, strategies);
        cb(buffer);

        opts.onCompleted.forEach(f => f(buffer));
        // or just
        // if (Array.isArray(opts.onCompleted)) opts.onCompleted.forEach(f => f(buffer));
    });
```

jhoffner commented:
@kazk What were the concerns with cleanup? We have started moving away from using temp folders because it often times makes it hard to reference other files. Since the container is one-time use we now just usually put everything inside of `'/home/codewarrior'`. See https://github.com/Codewars/codewars-runner-cli/blob/master/lib/shovel.js#L56. 

There is also a new [Github repo feature](https://github.com/Codewars/codewars-runner-cli/blob/master/documentation/setup_config_statements.md#download-github-repo) where you can pull in files from a github repo, which gets pulled into `opts.dir`, so using temp files would break that. 

If this came up as an issue, it might have been that you were testing out the CLI without going through Docker? If thats the workflow that you want to keep, maybe something that could be added is the ability to pass in "temp" as the `dir` parameter, which shovel would detect, create a temp folder and then set `opts.dir` to that new temp folder.

jhoffner commented:
Also for the additional packages step. Here are some suggestions that I pulled from [here](http://go-search.org/tops)

- sqlite3
- redis
- pq
- assert
- mock
- ghttp
- check
- [testify](https://github.com/stretchr/testify)

kazk commented:
Thanks for the suggestions. I'll look into them.

jhoffner commented:
> Does this mean you get a fresh environment for every it in go_spec.js?

No, its the same environment/process. I thought you meant cleanup after the entire suite ran. 

> I wanted to know how to use opts.onCompleted in shovel.start

This runs after the entire process has ran and the output has already been written. The only use of it right now is in [services.js](https://github.com/Codewars/codewars-runner-cli/blob/master/lib/services.js#L16). Its probably only going to be useful for you if you are having issues with cleanup while running mocha tests.

kazk commented:
> Currently there is no way of passing in onCompleted event handlers from within specs. You would have to do it from within go.js

That's what I tried, from `lib/runners/go.js`.

```javascript
// lib/runners/go.js
module.exports.run = function run(opts, cb) {
  if (!Array.isArray(opts.onCompleted)) opts.onCompleted = [];
  opts.onCompleted.push(function() {
    // cleanup here like in lib/services.js#L16
  });
  shovel.start(opts, cb, {
    solutionOnly: function (runCode) {
      runCode({name: 'go', args: ['run']});
    },
    testIntegration: function (runCode) {
      runCode({name: 'go', args: ['test']});
    }
  });
};
```

```javascript
// lib/shovel.js. shovel.start
module.exports.start = function start(opts, cb, strategies) {
  // ...
  opts.onCompleted = []; // this clears the passed in array from above (lib/shovel.js#L40)

  cb = cb || function (){};
  run(opts, strategies, function (buffer) {
    reportBuffer(opts, buffer, strategies);
    cb(buffer);
    opts.onCompleted.forEach(f => f(buffer));
  });
};
```

I just found`strategies.modifyOpts` [lib/shovel.js#L97](https://github.com/Codewars/codewars-runner-cli/blob/master/lib/shovel.js#L97). This might work?

```javascript
// lib/runners/go.js
module.exports.run = function run(opts, cb) {
  shovel.start(opts, cb, {
    solutionOnly: function (runCode) {
      runCode({name: 'go', args: ['run']});
    },
    testIntegration: function (runCode) {
      runCode({name: 'go', args: ['test']});
    },
    modifyOpts: function(opts) {
      opts.onCompleted.push(function() {
        // cleanup here like in lib/services.js#L16
      });
    }
  });
};
```

I ended up using `temp`, but I'll try this method if `temp` causes any problem.

kazk commented:
Additional package candidates. Needs feedback from Go users.

## Database

- sqlite3
  - Package sqlite3 provides interface to SQLite3 databases.
- redis
  - Package redis is a client for the Redis database.
- pq
  - Package pq is a pure Go Postgres driver for the database/sql package.

## Testing

- testify "github.com/stretchr/testify"
  - assert "github.com/stretchr/testify/assert"
    - Package assert provides a set of comprehensive testing tools for use with the normal Go testing system.
  - mock "github.com/stretchr/testify/mock"
    - Package mock provides a system by which it is possible to mock your objects and verify calls are happening as expected.

- check
  - Package check is a rich testing extension for Go's testing package.

- ghttp "github.com/onsi/gomega/ghttp"
  - Package ghttp supports testing HTTP clients by providing a test server (simply a thin wrapper around httptest's server) that supports registering multiple handlers.

jhoffner commented:
> I just foundstrategies.modifyOpts lib/shovel.js#L97. This might work?
Yeah that should do the trick. The code paths are kind of confusing. Its basically been transformed too many times over the years and has never had a full proper rewrite.

A few other packages possibly worth adding are some web frameworks. It hasn't taken off yet on CW but there is the ability already to test Rails/Express, etc. We hope to start pushing/encouraging this type of content and with the ghttp test library installed (and the new github download feature) it should be possible to write some.

kazk commented:
> The code paths are kind of confusing. Its basically been transformed too many times over the years and has never had a full proper rewrite.

Yeah, it took me a while to read/follow the code. I'm sure rewriting/documenting will lower the barrier of contributions to add new languages, but I've read your comment having a TODO list of 1k items.
Are you interested in contributions in those areas too? I've found some minor issues while studying the codebase. It's not much extra work for me because it'll be a byproduct of trying to understand how this works. If you're interested, I can go in deeper and propose changes.

---

Web framework kata will be interesting :)

jhoffner commented:
> Are you interested in contributions in those areas too?

Absolutely! Probably best to open it up as a separate PR.

kazk commented:
Ok, I'll do that :)

---

I have a question about outputting I forgot to ask. How is this message generated?

```
Time: 0ms Passed: 0 Failed: 0 Errors: 0
```

I'm guessing that Passed/Failed/Errors are just counting tags, but I'm not sure where to add the total time.

I'm enclosing the entire test suite with `<DESCRIBE::>` and `<COMPLETEDIN::>` and the output looks like [this](https://travis-ci.org/Codewars/codewars-runner-cli/builds/199004976#L617):

```
<DESCRIBE::>Kata Test
<DESCRIBE::>Add
<IT::>should add integers
<PASSED::>Test Passed
<COMPLETEDIN::>0.0467
<COMPLETEDIN::>
Passed: 1 Failed: 0
<COMPLETEDIN::>0.0962
PASS
ok  	_/tmp/go11716-11-1f8090c.mgabw97ldi	0.004s
```
(I'm going to remove that `Passed: 1 Failed: 0` line)

kazk commented:
I found that I can `console.log` the test outputs in kumite to see how it renders 😄 

![image](https://cloud.githubusercontent.com/assets/639336/22723199/534c8d7c-ed71-11e6-89c9-b7c97ce5ee1d.png)

---

Possible bug. When `<` is followed by letters (`[a-zA-Z]`), it disappears from the output.

So this one needs to be escaped.

```
<IT::>should add integers
<FAILED::>Test Failed
Expected<:LF:>    <int>: 0<:LF:>to equal<:LF:>    <int>: 2
<COMPLETEDIN::>0.4711
```
![image](https://cloud.githubusercontent.com/assets/639336/22723329/3e531fa2-ed72-11e6-8f09-0be64b336a7e.png)

But these are fine.
```
<IT::>should be empty
<FAILED::>Test Failed
Expected<:LF:>    <[]int | len:1, cap:1>: [1]<:LF:>to be empty
<COMPLETEDIN::>0.4555
```
![image](https://cloud.githubusercontent.com/assets/639336/22723361/776bcd8e-ed72-11e6-876d-07e0cea1359b.png)

```
<IT::>should have length of 2
<FAILED::>Test Failed
Expected<:LF:>    <[]int | len:1, cap:1>: [1]<:LF:>to have length 2
<COMPLETEDIN::>0.5768
```
![image](https://cloud.githubusercontent.com/assets/639336/22723370/9b30998e-ed72-11e6-852b-90e4e880fd6d.png)

jhoffner commented:
I see you figured out how to add tabs to log containers. 

> Possible bug. When < is followed by letters ([a-zA-Z]), it disappears from the output.

Yeah it's being interpreted as HTML, so escaping will do it. I've got to document the output format better, especially before I forget it all completely. I believe if you add a special modifier to an explicit log message you can also auto escape content without having to escape it yourself.  Try `<LOG:ESC:>${message}` and see if that does the trick.

> Time: 0ms Passed: 0 Failed: 0 Errors: 0

This is calculated automatically by counting all of the `<PASSED::>`, `<FAILED::>` and `<ERROR::>` lines.

kazk commented:
`<LOG:ESC:>` did it, but the first space is missing :/

![image](https://cloud.githubusercontent.com/assets/639336/22726638/6d27ccbc-ed89-11e6-9910-6e60397f7b1b.png)

```
<IT::>should add integers
<FAILED::>Test Failed
<LOG:ESC:>Expected<:LF:>    <int>: 0<:LF:>to equal<:LF:>    <int>: 2
<COMPLETEDIN::>0.4711

<IT::>should add integers
<FAILED::>Test Failed
<LOG:ESC:>Expected<:LF:>    <int>: 0<:LF:>to equal<:LF:>    <int>: 2
<COMPLETEDIN::>0.4711
```

---

This is off topic, but is 1 dan implemented? I finally reached "Overall:1 kyu / 100%" tonight and solved few katas since then so it should be over 100% now.

jhoffner commented:
> <LOG:ESC:> did it, but the first space is missing :/

Hmm thats odd. Its a rendering issue I can work on a fix from my end. 

> This is off topic, but is 1 dan implemented? I finally reached "Overall:1 kyu / 100%" tonight

Whoa! Congrats! Yeah it looks like there is a bug, you are currently in some kind of limbo state. It might be that you are actually at 99%, it looks like 80 more rank points are needed (which is about four 5 kyus). This all looks greek to me at the moment its been 3 years since I built it and I haven't had to touch it so I'll dig in and let you know.

kazk commented:
Thanks :)

> It might be that you are actually at 99%, it looks like 80 more rank points are needed

This makes sense because the chart does look like 99%.

![image](https://cloud.githubusercontent.com/assets/639336/22747124/f70e2dc6-edda-11e6-8474-228ef5bea9f8.png)

jhoffner commented:
Yeah it ended up just being an issue with the fact that we round percentages. I'll add a fix, though it looks like you went ahead and just solved the problem by getting to 100%! Congrats! 

Also looks like the black on black look isn't going to work, now that we actually have a 1 Dan on the site I can work on tweaking the colors.

kazk commented:
Thanks! It's been my secret goal to reach 1 dan first since someone told me that I was closest, so I'm happy :)

> Also looks like the black on black look isn't going to work

I really like black on black, but I agree that the ranking badge next to the name is hard to see.

kazk commented:
@jhoffner I think basic features are done.

Ginkgo and Gomega combined has much more feature and I don't know how far I should support and test. Also, I haven't looked into the new github downloading feature and I'm not sure how it works yet.

jhoffner commented:
@kazk really great work on this! I'm excited to have Go on the platform. 

One last thing that needs to be added, is a `examples/go.yml` file. There is also a test helper to verify that this file works. This will help me add examples to CW when I integrate.

kazk commented:
> @kazk really great work on this! I'm excited to have Go on the platform.

Thanks! I saw you opened #5 on June 10, 2014!!
I'm excited too, but little nervous because this is my first time adding a feature to a project with many users 😳 

I'll add `example/go.yml` and let you know.

kazk commented:
@jhoffner I added some examples.

I noticed that "bug fixes" example passed with incorrect answer so I added `<ERROR::>Build Failed` when `buffer.stdout` include '[build failed]'.

I had some trouble with YAML because it doesn't allow tab characters and standard coding convention in Go indents with tabs. Can you do something about it when inserting those examples in Codewars?

jhoffner commented on lib/runners/go.js:
```diff
@@ -53,6 +53,8 @@ func Test${name}(t *testing.T) {
         // FAIL	{dirname of test file}	{seconds}s
         sanitizeStdOut: function(stdout) {
+            // special case when build failed
+            if (stdout.includes('[build failed]')) return '<ERROR::>Build Failed\n';
```
Does any other information get written out when this happens? If so then maybe `stdout.replace('[build failed]', "\n<ERROR::>Build Failed\n")` would be better?

jhoffner replied:
BTW ignore my other "what is happening here" comments that you probably got notifications about.

kazk replied:
I think I omitted it because it gave out the answer, but I'll look into it and let you know.

kazk replied:
After '[build failed]', stdout contains the same information as stderr (last 3 lines are red).

```
FAIL	codewarrior/kata [build failed]
# codewarrior/kata
./kata.go:3: a + b evaluated but not used
./kata.go:4: missing return at end of function

      ✓ should have a passing bug fixes example (808ms)
# codewarrior/kata
./kata.go:3: a + b evaluated but not used
./kata.go:4: missing return at end of function
```
becomes
```
<ERROR::>Build Failed
# codewarrior/kata
./kata.go:3: a + b evaluated but not used
./kata.go:4: missing return at end of function
      1) should have a passing bug fixes example
```

jhoffner commented on lib/runners/go.js:
```diff
@@ -1,21 +1,74 @@
+        sanitizeStdOut: function(stdout) {
+            // special case when build failed
+            if (stdout.includes('[build failed]')) return '<ERROR::>Build Failed\n';
+            const ss = stdout.split('\n'), n = ss.length;
```
What exactly is happening here?

jhoffner commented on lib/runners/go.js:
```diff
@@ -1,21 +1,74 @@
+            const ss = stdout.split('\n'), n = ss.length;
+            for (let i = n - 1; i >= 0; --i) {
+                const s = ss[i];
+                if (s == 'PASS' || s.startsWith('--- FAIL'))
```
What exactly is happening here?

jhoffner commented:
@kazk Great job again. I have to say you impressed me with your thoroughness and attention to detail. I look forward to integrating this and getting it onto CW ASAP!

kazk commented:
Thanks for the compliment! Please let me know if you find any issues or have any questions :)

jhoffner commented:
@kazk preview.codewars.com has Go added. I wanted to give you a chance to test it out before I pushed to www. I did some testing and it looks great!

Just be careful not to publish anything yet until I push to www.

kazk commented:
I didn't expect you to do this in 4 days! Thanks for letting me know.

In the kata editor, is it possible indent the example code with tabs? I couldn't use tabs in go.yml because YAML doesn't allow it, but the standard convention is to use tabs to indent and spaces to align.
The standard coding conventions is followed much more commonly in Go compared to other languages because the tool is included. [go fmt your code](https://blog.golang.org/go-fmt-your-code) article from 2013 says 70% of the Go packages follows it, but I think it's much higher now.

jhoffner commented:
It looks like that might cause a refactoring effort, right now CW doesn't support tab characters and I have some custom indent/undent logic amongst other things that would have to be looked into. Probably is going to be better to just leave it. I can however adjust the characters (should it be 4 instead of 2 spaces?). 

Though its coding conventions for Go, usually tab conventions are really about keeping a code base uniform - so each language chooses either tabs or spaces. In CW's case, this wont be an issue since there is no choice - so I think it might be ok to break this type of convention. 

With that said IDK if it will cause issues with copy and pasting things from other places on the web (not that much copy and pasting should be going on anyway).

kazk commented:
Maybe indenting with tabs will be a problem on CodeWars.

I just tried writing some code indenting with tabs, but I found that the editor expands tabs when aligning the new line to previous line (Vim). It's confusing because the difference is not visible while editing.

![image](https://cloud.githubusercontent.com/assets/639336/22950388/47c137ca-f2ba-11e6-8979-bb6ee07777f4.png)

is

```go
func TwoOldestAges(ages []int) [2]int {
	a, b := 0, 0 // indented with tab
  for _, v := range ages { // automatically indented with two spaces
  	if v > b {
    	a, b = b, v
    } else if v > a {
    	a = v
    }
  }
  return [2]int{a, b}
}
```

![image](https://cloud.githubusercontent.com/assets/639336/22950780/a129b53e-f2bb-11e6-9dc9-8c5146ca7ca5.png)

tabs highlighted by searching `\t`.

---

I'll reply to the comment you just posted separately.

kazk commented:
> Probably is going to be better to just leave it.

Ok, I was just little worried that Go users will criticize that.

> I can however adjust the characters (should it be 4 instead of 2 spaces?).

I think `2` is fine.

> With that said IDK if it will cause issues with copy and pasting things from other places on the web (not that much copy and pasting should be going on anyway).

Copy and pasting shouldn't be a problem because it's not a syntax error to mix tabs and spaces.

jhoffner commented:
Cool, yeah for now lets leave it as is. 

In whatever time I get to myself I have been quietly working on an open source UI for interacting with Coderunner, which will hopefully eventually replace all of the IDE UI's on CW. We can look at supporting tabs in there.

kazk commented:
That sounds great 👍  I'll look forward to it :)

I haven't noticed any other concerns in kata editor or kumite.
